### PR TITLE
Working with Ktor-Client: How to Build and Make HTTP Requests With Ktor-Client in Android Jetpack Compose

### LINKED ISSUE #4858: Working with Ktor-Client: How to Build and Make HTTP Requests With Ktor-Client in Android
#### Proposed title of article

Working with Ktor-Client: Build Simple REST API to Make HTTP Requests With Ktor-Client

#### Proposed article introduction

Ktor is a server-side framework that helps us build applications in kotlin. It is a modern framework which means it is an asynchronous framework built on top of kotlin coroutines. It is not only a back-end framework with kotlin but also a client, so a network library such as okhttp and Retrofit. We can use it to make network requests HTTP requests to an API to get the response back to an application. This will learn more about Ktor. We will set up Ktor-Client to make HTTP Requests And finally build a simple REST API with Ktor.

#### Key takeaways 

- Ktor architecture
- Ktor features
- Ktor vs Retrofit
- Build and Make HTTP Requests With Ktor-Client Android

#### Article quality
How is your article unique? Tell us what makes your approach different from similar articles that have published on the same topic?
Is yours more in-depth? Does it cover additional topics? Do you provide handy tips or anecdotal advice?

#### References
Please list links to any published content/research that you intend to use to support/guide this article.

### Conclusion
Finally, remove the Pre-Submission advice section and all our blockquoted notes as you fill in the form before you submit.
We look forwarding to reviewing your topic suggestion.

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by manmusa100**
### Article checklist

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

This closes #4858

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by louisefindlay23** [comment]
Plagiarism check complete

File: content/articles/http-requests-with-ktor-client-in-android-jetpack-compose/index.md [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+
+### Adding Jetpack Compose UI
+We have the data instance ready, we have processed all the requests and responses to the API server. Now we need to populate this returned data into some Jetpack Compose components.
+
+Typically to display a large number of items in an android application, you would set up a Recycler View adapter. With Jetpack Compose, you don't need to set up this adapter since you won't use any XML UIs. Jetpack Compose allows you to use `LazyColumn` that replaces the typical Recycler View adapter. You don't even need a View holder class since Jetpack Compose allows you to achieve the same thing we just a few lines of code. Check this guide to learn more about the [Jetpack Compose LazyColumn](/engineering-education/building-scrollable-and-lazy-components-in-jetpack-compose/)
+
+Lets we how we can set up this list of products using the Jetpack Compose `LazyColumn`. `LazyColumn` allows you to display list items in a vertical orientation. And if you want to achieve the horizontal orientation, you use `LazyRow`.
+
```
Comment (Ericgacoki) [inline content/articles/http-requests-with-ktor-client-in-android-jetpack-compose/index.md:259]: Did you mean `Let's see how we can...`?
If so, fix the first sentence.

File: content/articles/http-requests-with-ktor-client-in-android-jetpack-compose/index.md [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+This will create a new instance of the Lazy that uses the specified initialization function initializer and the default thread-safety mode `LazyThreadSafetyMode.SYNCHRONIZED`.
+
+Create an empty `initialValue` that returns an observable snapshot State that produces values over time without a defined data source. This will allow the client to access the data source asynchronously as data is being populated to the set Jetpack Compose states. Also, we will add a `producer` that returns the values of the defined data source. Add the following code block inside the application Theme block.
+
+```kotlin
+val products = produceState(
+    initialValue = emptyList<ResponseModel>(),
+    producer = {
+        value = apiService.getProducts()
```
Comment (Ericgacoki) [inline content/articles/http-requests-with-ktor-client-in-android-jetpack-compose/index.md:276]: 👏

File: content/articles/http-requests-with-ktor-client-in-android-jetpack-compose/index.md [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+This column represents an actual Recycler View. Here we are using the Jetpack Compose to populate the data to a cleaned and minimalist UI with less code.
+
+We are wrapping each item in a `Box` that has three properties, the `Image` and two `Text` blocks.
+
+First, we will process the returned data and get the value of the `image`. The returned string is a URL that loads an image from the server. To load these images, we are using a Coil image loader. Here we set the image URL, pass the parameter `data` and assign it the value of the returned image URL. This will launch an image request. If return `true`, we proceed with the request and load the image into the Image view using the Painter. This will create a composable that lays out and draws a given Painter. This will attempt to size the composable according to the Painter's intrinsic size and the `Modifier` parameters.
+
+If it returns `false`, we skip executing the request and builder an optional lambda) configure the request, and set an error drawable resources. The Text composable will load the values of product `title` and `description`.
+
+The app is now ready, and you can run it to test if everything works as expected.
+
```
Comment (Ericgacoki) [inline content/articles/http-requests-with-ktor-client-in-android-jetpack-compose/index.md:348]: Just before the conclusion, provide at least one image - a screenshot of what to expect upon running the app.

**Review by Ericgacoki — CHANGES_REQUESTED** [review]
Hello @manmusa100,
It's amazing to see how well you've covered Ktor in this tutorial. Just a few changes (see the comments attached below) and we'll be good to go :rocket:

**Comment by manmusa100** [comment]
Hi @Ericgacoki, Check the changes I have added.

`gradle` identifier failed to render the code syntax, I have used `koltin` instead.

**Comment by Ericgacoki** [comment]
> Hi @Ericgacoki, Check the changes I have added.
> 
> `gradle` identifier failed to render the code syntax, I have used `koltin` instead.

`gradle` usually doesn't highlight the syntax properly especially on VsCode. Don't worry about it 😇

**Review by Ericgacoki — APPROVED** [review]

**Comment by LinusMuema** [comment]
Hi @manmusa100 

Great work on the article.

cc: @Ericgacoki

**Comment by hectorkambow** [comment]
@manmusa100 I think this article has the wrong author on it 🤔  

@Ericgacoki 
cc @LinusMuema 

not sure - trying to double check

**Comment by Ericgacoki** [comment]
I've checked @manmusa100's author file(s) and his previous article(s). The author name is similar to the one used in this article (`author: moses-chege`).

**Pull request merged**